### ISSUE TITLE
Basis needs to have plug and minus operators added

### ISSUE THREAD

**Issue by blackears** [godotengine/godot-proposals#8633]
### Describe the project you are working on

A math library which to performs matrix decompisitions.

### Describe the problem or limitation you are having in your project

Godot lacks a proper library for linear algebra.  Instead it provides us with classes like Basis and Transform3D which are specialized for 3D rendering.  Which makes things difficult if you want to use these matrices for more general mathematics.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Since the Basis class is essentially a 3x3 matrix, it would be useful to add plus and minus operators to it so it can be used for general 3D matrix operations and not just specialized for the 3D render pipeline.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add `operator +(Basis right)` and `operator -(Basis right)` operators to the Basis class.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, but it is inconvenient and makes the code harder to read.

### Is there a reason why this should be core and not an add-on in the asset library?

This is an essential operation I would think most would expect to be part of a 3D math library.

**Comment by AThousandShips** [comment]
Why just `Basis`? `Transform2D/3D` are also just matrices, fully accessible as such

**Comment by blackears** [comment]
Transform2D and Transform3D have the convention where the fourth row is omitted - ie, always presumed to be (0, 0, 0, 1) which is true for what you usually use them for transforming meshes.  So a Transform3D is really a 4D matrix where the fourth row is implied.  A Transform2D is a 3D matrix where the third row is implied.

This is a problem when you're using the matrices for calculations and not just transforms in 3D space.  I'm using a Basis because I need a 3D matrix with a third row.  I'm trying to decompose a matrix into a translation, scale, rotation and shear component.  The shear component is not part of the Godot library, so I need to write it myself which requires a bit of advanced math.  In particular, I need to subtract one matrix from another.

**Comment by AThousandShips** [comment]
But `Basis` isn't a matrix, it is *represented* as a matrix, in the exact same sense that `Transform`s are, and adding `Basis` to a nother doesn't make sense for their intended use, AFAIK

**Comment by blackears** [comment]
Well, I need a matrix.  Does Godot have those?  Otherwise I have to write my own matrix library.

**Comment by AThousandShips** [comment]
Well that's a bit of a separate proposal I'd say, and exists:
* https://github.com/godotengine/godot-proposals/issues/5148

I'd say if there's not a specific reason for `Basis`, when used as one, to have these operators bound it shouldn't

**Comment by blackears** [comment]
I agree it would be a good idea.  The proposal here seems to be more for a general purpose multi dimensional grid.  I need a numerical matrix intended for math operations.  I agree that I am probably abusing the Basis class here, but a mathematical matrix library would be a good addition to Godot, in my opinion.

**Comment by AThousandShips** [comment]
I think there's another proposal or two out there too, for matrix in particular

**Comment by blackears** [comment]
Is this something the ought to be implemented in C++?

**Comment by AThousandShips** [comment]
It'd be best to do so yeah, essentially everything in GDScript has a c++ backing, the question is however how generally needed matrices would be

**Comment by blackears** [comment]
Well, for people doing more advanced stuff - especially if you're trying to get developers into using compute shaders.  Not so much for the average user.  But such a library should be pretty small and easy to implement.

**Comment by AThousandShips** [comment]
I think that since:
* There's a very trivial workaround that doesn't hurt readability or confusees, namely:
```gdscript
func add_basis(lhs: Basis, rhs: Basis) -> Basis:
    return Basis(lhs.x + rhs.x, lhs.y + rhs.y, lhs.z + rhs.z)

func sub_basis(lhs: Basis, rhs: Basis) -> Basis:
    return Basis(lhs.x - rhs.x, lhs.y - rhs.y, lhs.z - rhs.z)
```
* There's no real use case for adding two transformation matrices together like this, so it's only useful when treating `Basis` as a general matrix

I'd say this should be closed in favor of a dedicated generic matrix proposal, which I'd say would be better suited as an extension or a plugin as it's of limited use in general games

**Comment by blackears** [comment]
That raises another issue - at the moment it's not easy in Godot to use more than one addon at a time due to potential class name conflicts.  And the Godot Asset Library has a restriction that your addon cannot depend on any other addons.
